File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *first = head, *curr = head, *prev = nullptr;
        int i = n;
        while(first) {
            if(i <= 0) {
                prev = curr;
                curr = curr->next;
            }
            i--;
            first = first->next;
        }
        
        // remove curr
        if(prev) {
            prev->next = curr->next;
            curr->next = nullptr;
        } else { // prev == nullptr, curr == head
            head = head->next;
        }
        
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        prev, curr, post = None, head, head
        for i in range(n):
            post = post.next
        while post:
            prev, curr, post = curr, curr.next, post.next
        if prev:
            prev.next = curr.next
        else: # curr == head
            head = head.next
        
        return head